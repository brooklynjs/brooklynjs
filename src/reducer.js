import { nextEventIndex } from './utils/events';

// Initial State
export const initialState = {
  allEvents: [],
  currentEvent: {
    eventDate: null,
    speakers: [],
    sponsors: []
  },
  currentEventIndex: null
};

// Actions
export const initialLoad = (events, now = new Date()) => ({
  type: 'INITIAL_LOAD', events, now
});

export const loadNextEvent = () => ({
  type: 'LOAD_NEXT_EVENT'
})

export const loadPreviousEvent = () => ({
  type: 'LOAD_PREVIOUS_EVENT'
});

// Reducer

const reducer = (state, action) => {
  switch(action.type) {
    case 'INITIAL_LOAD':
      return handleInitialLoad(state, action);

    case 'LOAD_NEXT_EVENT':
      return handleLoadNextEvent(state, action);

    case 'LOAD_PREVIOUS_EVENT':
      return handleLoadPreviousEvent(state, action);

    default:
      return state;
  }
};

const handleInitialLoad = (state, action) => {
  const { events, now } = action;
  const eventIndex = nextEventIndex(events, now);

  return {
    ...state,
    allEvents: events,
    currentEvent: events[eventIndex],
    currentEventIndex: eventIndex
  };
};

const handleLoadNextEvent = (state, action) => {
  const { allEvents, currentEventIndex } = state;

  if (currentEventIndex + 1 < allEvents.length) {
    const nextEventIndex = currentEventIndex + 1;
    const nextEvent = allEvents[nextEventIndex];

    return {
      ...state,
      currentEvent: nextEvent,
      currentEventIndex: nextEventIndex
    }
  } else {
    return state;
  }
};

const handleLoadPreviousEvent = (state, action) => {
  const { allEvents, currentEventIndex } = state;

  if (currentEventIndex > 0) {
    const previousEventIndex = currentEventIndex - 1;
    const previousEvent = allEvents[previousEventIndex];

    return {
      ...state,
      currentEvent: previousEvent,
      currentEventIndex: previousEventIndex
    };
  } else {
    return state;
  }
};

export default reducer;
