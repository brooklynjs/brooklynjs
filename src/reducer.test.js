import reducer, {
  initialLoad,
  initialState,
  loadNextEvent,
  loadPreviousEvent
} from './reducer';

import {
  events,
  futureEvent,
  previousEvent,
  today,
  todayEvent
} from './fixtures';

describe('reducer', () => {
  describe('INITIAL_LOAD', () => {
    it('loads the event data and selects the next event', () => {
      const newState = reducer(initialState, initialLoad(events, today));

      expect(newState).toMatchObject({
        allEvents: events,
        currentEvent: todayEvent,
        currentEventIndex: 1
      });

      expect(newState).toMatchSnapshot();
    });
  });

  describe('LOAD_NEXT_EVENT', () => {
    describe('when the currentEvent is not the last event', () => {
      const state = {
        ...initialState,
        allEvents: events,
        currentEvent: todayEvent,
        currentEventIndex: 1
      };

      it('sets the current event to be the next event', () => {
        const newState = reducer(state, loadNextEvent());

        expect(newState).toMatchObject({
          currentEvent: futureEvent,
          currentEventIndex: 2
        });

        expect(newState).toMatchSnapshot();
      });
    });

    describe('when the currentEvent is already the last event', () => {
      const state = {
        ...initialState,
        allEvents: events,
        currentEvent: futureEvent,
        currentEventIndex: 2
      };

      it('does not modify the state', () => {
        const newState = reducer(state, loadNextEvent());

        expect(newState).toEqual(state);
      });
    });
  });

  describe('LOAD_PREVIOUS_EVENT', () => {
    describe('when the currentEvent is not the first event', () => {
      const state = {
        ...initialState,
        allEvents: events,
        currentEvent: futureEvent,
        currentEventIndex: 2
      };

      it('sets the current event to be the previous event', () => {
        const newState = reducer(state, loadPreviousEvent());

        expect(newState).toMatchObject({
          currentEvent: todayEvent,
          currentEventIndex: 1
        });

        expect(newState).toMatchSnapshot();
      });
    });

    describe('when the currentEvent is the first event', () => {
      const state = {
        ...initialState,
        allEvents: events,
        currentEvent: previousEvent,
        currentEventIndex: 0
      };

      it('does not modify the state', () => {
        const newState = reducer(state, loadPreviousEvent());

        expect(newState).toEqual(state);
      });
    });
  });
});
