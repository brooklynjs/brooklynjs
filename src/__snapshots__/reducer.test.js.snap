// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducer INITIAL_LOAD loads the event data and selects the next event 1`] = `
Object {
  "allEvents": Array [
    Object {
      "eventDate": 2017-05-21T03:59:59.999Z,
      "speakers": Array [
        Object {
          "name": "Awesome Person",
          "title": "Awesome Talk",
          "url": "http://www.example.com/awesome",
        },
        Object {
          "name": "Chill Person",
          "title": "Chill Talk",
          "url": "http://www.example.com/chill",
        },
        Object {
          "name": "Dope Person",
          "title": "Dope Talk",
          "url": "http://www.example.com/dope",
        },
        Object {
          "name": "Righteous Person",
          "title": "Righteous Talk",
          "url": "http://www.example.com/righteous",
        },
        Object {
          "name": "Wild Person",
          "title": "Wild Talk",
          "url": "http://www.example.com/wild",
        },
      ],
      "sponsors": Array [
        Object {
          "name": "Generous Sponsor",
          "url": "http://www.example.com/generous",
        },
      ],
    },
    Object {
      "eventDate": 2017-09-22T03:59:59.999Z,
      "speakers": Array [
        Object {
          "name": "Chill Person",
          "title": "Chill Talk",
          "url": "http://www.example.com/chill",
        },
        Object {
          "name": "Dope Person",
          "title": "Dope Talk",
          "url": "http://www.example.com/dope",
        },
      ],
      "sponsors": Array [],
    },
    Object {
      "eventDate": 2017-10-19T04:00:00.000Z,
      "speakers": Array [],
      "sponsors": Array [],
    },
  ],
  "currentEvent": Object {
    "eventDate": 2017-09-22T03:59:59.999Z,
    "speakers": Array [
      Object {
        "name": "Chill Person",
        "title": "Chill Talk",
        "url": "http://www.example.com/chill",
      },
      Object {
        "name": "Dope Person",
        "title": "Dope Talk",
        "url": "http://www.example.com/dope",
      },
    ],
    "sponsors": Array [],
  },
  "currentEventIndex": 1,
}
`;

exports[`reducer LOAD_NEXT_EVENT when the currentEvent is not the last event sets the current event to be the next event 1`] = `
Object {
  "allEvents": Array [
    Object {
      "eventDate": 2017-05-21T03:59:59.999Z,
      "speakers": Array [
        Object {
          "name": "Awesome Person",
          "title": "Awesome Talk",
          "url": "http://www.example.com/awesome",
        },
        Object {
          "name": "Chill Person",
          "title": "Chill Talk",
          "url": "http://www.example.com/chill",
        },
        Object {
          "name": "Dope Person",
          "title": "Dope Talk",
          "url": "http://www.example.com/dope",
        },
        Object {
          "name": "Righteous Person",
          "title": "Righteous Talk",
          "url": "http://www.example.com/righteous",
        },
        Object {
          "name": "Wild Person",
          "title": "Wild Talk",
          "url": "http://www.example.com/wild",
        },
      ],
      "sponsors": Array [
        Object {
          "name": "Generous Sponsor",
          "url": "http://www.example.com/generous",
        },
      ],
    },
    Object {
      "eventDate": 2017-09-22T03:59:59.999Z,
      "speakers": Array [
        Object {
          "name": "Chill Person",
          "title": "Chill Talk",
          "url": "http://www.example.com/chill",
        },
        Object {
          "name": "Dope Person",
          "title": "Dope Talk",
          "url": "http://www.example.com/dope",
        },
      ],
      "sponsors": Array [],
    },
    Object {
      "eventDate": 2017-10-19T04:00:00.000Z,
      "speakers": Array [],
      "sponsors": Array [],
    },
  ],
  "currentEvent": Object {
    "eventDate": 2017-10-19T04:00:00.000Z,
    "speakers": Array [],
    "sponsors": Array [],
  },
  "currentEventIndex": 2,
}
`;

exports[`reducer LOAD_PREVIOUS_EVENT when the currentEvent is not the first event sets the current event to be the previous event 1`] = `
Object {
  "allEvents": Array [
    Object {
      "eventDate": 2017-05-21T03:59:59.999Z,
      "speakers": Array [
        Object {
          "name": "Awesome Person",
          "title": "Awesome Talk",
          "url": "http://www.example.com/awesome",
        },
        Object {
          "name": "Chill Person",
          "title": "Chill Talk",
          "url": "http://www.example.com/chill",
        },
        Object {
          "name": "Dope Person",
          "title": "Dope Talk",
          "url": "http://www.example.com/dope",
        },
        Object {
          "name": "Righteous Person",
          "title": "Righteous Talk",
          "url": "http://www.example.com/righteous",
        },
        Object {
          "name": "Wild Person",
          "title": "Wild Talk",
          "url": "http://www.example.com/wild",
        },
      ],
      "sponsors": Array [
        Object {
          "name": "Generous Sponsor",
          "url": "http://www.example.com/generous",
        },
      ],
    },
    Object {
      "eventDate": 2017-09-22T03:59:59.999Z,
      "speakers": Array [
        Object {
          "name": "Chill Person",
          "title": "Chill Talk",
          "url": "http://www.example.com/chill",
        },
        Object {
          "name": "Dope Person",
          "title": "Dope Talk",
          "url": "http://www.example.com/dope",
        },
      ],
      "sponsors": Array [],
    },
    Object {
      "eventDate": 2017-10-19T04:00:00.000Z,
      "speakers": Array [],
      "sponsors": Array [],
    },
  ],
  "currentEvent": Object {
    "eventDate": 2017-09-22T03:59:59.999Z,
    "speakers": Array [
      Object {
        "name": "Chill Person",
        "title": "Chill Talk",
        "url": "http://www.example.com/chill",
      },
      Object {
        "name": "Dope Person",
        "title": "Dope Talk",
        "url": "http://www.example.com/dope",
      },
    ],
    "sponsors": Array [],
  },
  "currentEventIndex": 1,
}
`;
